pkg/service/controller/servingcert/controller.go

```diff
@@ -0,0 +1,232 @@
+func (sc *ServiceServingCertController) syncService(key string) error {
+	obj, exists, err := sc.serviceCache.GetByKey(key)
+	if err != nil {
+		glog.Infof("Unable to retrieve service %v from store: %v", key, err)
+		sc.queue.Add(key)
+		return err
+	}
+	if !exists {
+		glog.Infof("Service has been deleted %v", key)
+		return nil
+	}
+
+	// make a copy to avoid mutating cache state
+	t := *obj.(*kapi.Service)
+	service := &t
+
+	if !sc.requiresCertGeneration(service) {
+		return nil
+	}
+
+	dnsName := service.Name + "." + service.Namespace + ".svc"
+	fqDNSName := dnsName + "." + sc.dnsSuffix
+	servingCert, err := sc.ca.MakeServerCert(sets.NewString(dnsName, fqDNSName))
+	if err != nil {
+		return err
+	}
+	certBytes, keyBytes, err := servingCert.GetPEMBytes()
+	if err != nil {
+		return err
+	}
+
+	secret := &kapi.Secret{}
+	secret.Name = service.Annotations[ServingCertSecretAnnotation]
+	secret.Namespace = service.Namespace
+	secret.Type = kapi.SecretTypeTLS
+	secret.Data = map[string][]byte{}
+	secret.Data[kapi.TLSCertKey] = certBytes
+	secret.Data[kapi.TLSPrivateKeyKey] = keyBytes
+	secret.Annotations = map[string]string{}
+	secret.Annotations[ServiceUIDAnnotation] = string(service.UID)
+	secret.Annotations[ServiceNameAnnotation] = service.Name
+	_, err = sc.secretClient.Secrets(service.Namespace).Create(secret)
+	if err != nil && !kapierrors.IsAlreadyExists(err) {
+		// if we have an error creating the secret, then try to update the service with that information.  If it fails,
+		// then we'll just try again later on  re-list or because the service had already been updated and we'll get triggered again.
+		service.Annotations[ServingCertErrorAnnotation] = err.Error()
+		sc.serviceClient.Services(service.Namespace).Update(service)
+		return err
+	}
+	if kapierrors.IsAlreadyExists(err) {
+		actualSecret, err := sc.secretClient.Secrets(service.Namespace).Get(secret.Name)
+		if err != nil {
+			// if we have an error creating the secret, then try to update the service with that information.  If it fails,
+			// then we'll just try again later on  re-list or because the service had already been updated and we'll get triggered again.
+			service.Annotations[ServingCertErrorAnnotation] = err.Error()
+			sc.serviceClient.Services(service.Namespace).Update(service)
+			return err
+		}
+
+		if actualSecret.Annotations[ServiceUIDAnnotation] != string(service.UID) {
+			service.Annotations[ServingCertErrorAnnotation] = fmt.Sprintf("secret/%v references serviceUID %v, which does not match %v", actualSecret.Name, actualSecret.Annotations[ServiceUIDAnnotation], service.UID)
+			sc.serviceClient.Services(service.Namespace).Update(service)
+			return errors.New(service.Annotations[ServingCertErrorAnnotation])
+		}
+	}
+
+	service.Annotations[ServingCertCreatedByAnnotation] = sc.ca.Config.Certs[0].Subject.CommonName
+	_, err = sc.serviceClient.Services(service.Namespace).Update(service)
```

Comment:
> Make sure that we cleanup the error annotation

Added key deletion, but the current code doesn't try to gen keys if a previous error occurred, so its not strictly need.  Prevents rot though.